cap-primitives/src/yanix/linux/fs/open.rs

```diff
@@ -35,57 +35,81 @@ struct OpenHow {
 }
 const SIZEOF_OPEN_HOW: usize = std::mem::size_of::<OpenHow>();
 
-/// Call the `openat2` system call.
-fn open_openat2(start: &fs::File, path: &Path, options: &OpenOptions) -> io::Result<fs::File> {
-    let oflags = compute_oflags(options);
-    let mode = options.ext.mode;
+/// Call the `openat2` system call. If the syscall is unavailable, mark it so for future
+/// calls, and fallback to `open_manually_wrapper`
+fn openat2_or_open_manually(
+    start: &fs::File,
+    path: &Path,
+    options: &OpenOptions,
+) -> io::Result<fs::File> {
+    static INVALID: AtomicBool = AtomicBool::new(false);
+    if !INVALID.load(SeqCst) {
+        let oflags = compute_oflags(options);
+        let mode = options.ext.mode;
 
-    let path_cstr = CString::new(path.as_os_str().as_bytes())?;
-    let open_how = OpenHow {
-        oflag: u64::from(oflags.bits() as u32),
-        mode: u64::from(mode),
-        resolve: RESOLVE_BENEATH | RESOLVE_NO_MAGICLINKS,
-    };
+        // Check for empty path, and if empty, change to ".".
+        let path = if path == Path::new("") {
+            &Path::new(".")
+        } else {
+            path
+        };
+        let path_cstr = CString::new(path.as_os_str().as_bytes())?;
+        let open_how = OpenHow {
+            oflag: u64::from(oflags.bits() as u32),
+            mode: u64::from(mode),
+            resolve: RESOLVE_BENEATH | RESOLVE_NO_MAGICLINKS,
+        };
+        // `openat2` fails with `EAGAIN` if a rename happens anywhere on the host
+        // while it's running, so use a loop to retry it a few times. But not too many
+        // times, because there's no limit on how often this can happen.
+        for _ in 0..4 {
+            unsafe {
+                match libc::syscall(
+                    SYS_OPENAT2,
+                    start.as_raw_fd(),
+                    path_cstr.as_ptr(),
+                    &open_how,
+                    SIZEOF_OPEN_HOW,
+                ) {
+                    -1 => match io::Error::last_os_error().raw_os_error().unwrap() {
+                        libc::EAGAIN => continue,
+                        libc::EXDEV => {
+                            if path.is_absolute() {
+                                return absolute_path();
+                            }
```

Comment:
This unfortunately doesn't handle the case where `path` is a symlink to an absolute path, so it'll still print a different message than `open_manually` in that case. I like the overall direction of this patch, moving toward making the error messages consistent between the different implementations, so it seems like we should take it even further and just have one message for both ".." and "/" escape attempts.

Reply:
Makes sense, thanks!